File: src/adapters/pi/scan.ts
import fs from 'node:fs';
import path from 'node:path';

import type {
  AdapterConfig,
  AdapterSnapshot,
  CategoryConfig,
  CategorySnapshot,
  SnapshotEntry,
  SnapshotFiles,
} from '../../core/types.js';
import { entryKindFor } from '../../core/entry-kind.js';
import { expandHome } from '../../core/paths.js';
import { matchesIgnore } from './ignore.js';

export interface ScanError {
  path: string;
  message: string;
}

export interface ScanOutcome {
  snapshot: AdapterSnapshot; // 仅含 root 存在且 enabled 的分类
  errors: ScanError[]; // root 不存在 → snapshot 空 + error；读失败不炸
}

/** 目录型 path（以 '/' 结尾）→ true。 */
function isDirPath(p: string): boolean {
  return p.endsWith('/');
}

function normalizeRel(rel: string): string {
  return rel.split(path.sep).join('/');
}

function basenameOf(rel: string): string {
  const idx = rel.lastIndexOf('/');
  return idx === -1 ? rel : rel.slice(idx + 1);
}

/**
 * category 内 exclude 判定：文件名 glob。
 * 语义 = matchesIgnore 应用于 relPath、其 basename、以及每一段路径名。
 * （`*cache*` 因此既排除 `foo-cache.js`，也排除 `cache/`、`mycache-tool/index.ts` 这类目录整棵。）
 */
function isExcluded(relPath: string, exclude: string[] | undefined): boolean {
  if (!exclude || exclude.length === 0) return false;
  if (matchesIgnore(relPath, exclude)) return true;
  if (matchesIgnore(basenameOf(relPath), exclude)) return true;
  const segments = relPath.split('/').filter((s) => s.length > 0);
  return segments.some((segment) => matchesIgnore(segment, exclude));
}

/**
 * adapter 级 ignore：§1.5 冻结为“相对 root 的路径 glob”。
 * 严格按 relPath 匹配（不做 basename / 分段兜底），避免 glob 语义膨胀；
 * 目录前缀模式（`sessions/`）天然覆盖整棵子树，因为 walk 逐层检查每个 relPath。
 */
function isIgnored(relPath: string, ignore: string[] | undefined): boolean {
  if (!ignore || ignore.length === 0) return false;
  return matchesIgnore(relPath, ignore);
}

const MAX_DEPTH = 32;

interface WalkState {
  ignore: string[] | undefined;
  exclude: string[] | undefined;
  /** symlink 逃逸 allowlist（相对 adapter root 的 glob，§2.0-1 / D7）。 */
  allowEscape: string[] | undefined;
  errors: ScanError[];
  out: string[];
  /** adapter root 的真实路径（symlink containment 基准）。 */
  rootReal: string;
  /**
   * allowlist 匹配基准：本次 walk 起点相对 **adapter root** 的路径（= 声明的 category path）。
   * 快照 key（`rel`）是相对 **category 目录** 的，而 allowEscape 是相对 root 的 glob，
   * 故匹配时必须用 `joinRel(rootRelBase, rel)` 还原成 root 相对路径。
   */
  rootRelBase: string;
  /** 本次 walk 已展开过的目录真实路径（symlink 回环检测）。 */
  visited: Set<string>;
}

/** 拼接 root 相对路径（rootRelBase 与 category 内 rel；两段都可能为空）。 */
function joinRel(rootRelBase: string, rel: string): string {
  if (rootRelBase === '') return rel;
  if (rel === '') return rootRelBase;
  return `${rootRelBase}/${rel}`;
}

/**
 * 逃逸是否被 allowlist 放行（D7）：对 **root 相对路径** 跑 `matchesIgnore`，语义与
 * adapter 级 ignore 完全一致（字面量 + `*` 不跨 `/`，尾 `/` 为目录前缀）；
 * 缺省 / 空数组 = 维持 M1 安全边界。
 */
function isEscapeAllowed(rootRel: string, allowEscape: string[] | undefined): boolean {
  if (!allowEscape || allowEscape.length === 0) return false;
  return matchesIgnore(rootRel, allowEscape);
}

/** realpath 封装：失败（悬空 / 竞争删除）→ undefined。 */
function tryRealpath(target: string): string | undefined {
  try {
    return fs.realpathSync(target);
  } catch {
    return undefined;
  }
}

/** child 是否位于 root 之下（两者都必须是 realpath）。 */
function isWithinRoot(child: string, root: string): boolean {
  if (child === root) return true;
  const prefix = root.endsWith(path.sep) ? root : `${root}${path.sep}`;
  return child.startsWith(prefix);
}

/**
 * symlink 条目处理（安全边界）：
 *   1. realpath 失败（悬空链接）→ 静默跳过（等价于「该路径不存在」）；
 *   2. realpath 落在 adapter root 之外 → skip + 记 ScanError（防「链接把 root 外内容带进快照」），
 *      **除非** root 相对路径命中 `allowEscape`（D7，显式 opt-in）→ 视同 root 内链接继续走 3/4；
 *   3. realpath 指向目录 → 交给 walk（其入口用 visited 集合截断回环）；
 *   4. realpath 指向文件 → 正常收集。
 *
 * 注意 allowlist 只放行**这条链接本身**：逃逸目录内部的后续 symlink 仍按同规则重新判定
 * （要连带放行整棵子树，用尾 `/` 前缀模式，如 `skills/agent-browser/`）。
 */
function visitSymlink(abs: string, rel: string, depth: number, state: WalkState): void {
  const real = tryRealpath(abs);
  if (real === undefined) return;

  if (!isWithinRoot(real, state.rootReal) && !isEscapeAllowed(joinRel(state.rootRelBase, rel), state.allowEscape)) {
    state.errors.push({
      path: abs,
      message: `symlink 逃逸 adapter root: ${real}`,
    });
    return;
  }

  let st: fs.Stats;
  try {
    st = fs.statSync(real);
  } catch {
    return;
  }
  if (st.isDirectory()) {
    walk(abs, rel, depth + 1, state);
  } else if (st.isFile()) {
    state.out.push(rel);
  }
}

/**
 * 递归收集目录下所有文件的 relPath（相对扫描起点），已应用 ignore。
 * 目录本身命中 ignore / exclude 则整棵剪枝。
 *
 * 顺序：readdir 原始顺序（不再在内层排序——scanCategory 末尾的统一排序才是稳定顺序的唯一来源，
 * 两处排序纯属冗余）。symlink 见 visitSymlink：目录真实路径记入 visited，回环只展开一次。
 */
function walk(absBase: string, relBase: string, depth: number, state: WalkState): void {
  if (depth > MAX_DEPTH) return;

  const realBase = tryRealpath(absBase);
  if (realBase === undefined) return;
  if (state.visited.has(realBase)) return; // symlink 目录回环：不再重复展开
  state.visited.add(realBase);

  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(absBase, { withFileTypes: true });
  } catch (err) {
    state.errors.push({ path: absBase, message: (err as Error).message });
    return;
  }
  for (const entry of entries) {
    const rel = relBase === '' ? entry.name : `${relBase}/${entry.name}`;
    if (isIgnored(rel, state.ignore) || isExcluded(rel, state.exclude)) continue;
    const abs = path.join(absBase, entry.name);

    if (entry.isSymbolicLink()) {
      visitSymlink(abs, rel, depth, state);
      continue;
    }
    if (entry.isDirectory()) {
      walk(abs, rel, depth + 1, state);
    } else if (entry.isFile()) {
      state.out.push(rel);
    }
  }
}

function readEntry(abs: string, mode: CategoryConfig['mode'], errors: ScanError[]): SnapshotEntry | undefined {
  let content: string;
  try {
    content = fs.readFileSync(abs, 'utf8');
  } catch (err) {
    errors.push({ path: abs, message: (err as Error).message });
    return undefined;
  }
  return { kind: entryKindFor(mode, content), content };
}

/**
 * 解析用户在 homer.json 里声明的 path（单文件 / 目录）——同样适用 root containment。
 *
 * 计划原文：「对每个 symlink 先 fs.realpathSync 解析真实路径——realpath 不在 root realpath 之下
 * → skip + 记 ScanError（防逃逸）」。因此**包括**声明的 category path 自身为 symlink 的情况。
 * 返回 undefined = 该 path 不存在（缺文件/缺目录，不视为错误）。
 *
 * D7（§2.0-1）：与 `visitSymlink` 共用同一逃逸判定 —— 声明的 path 逃逸时同样先查
 * `allowEscape`（匹配基准 = 声明的 path 去掉尾 `/` 后的 root 相对路径），命中才跟随。
 */
function resolveConfiguredPath(
  abs: string,
  rootReal: string,
  errors: ScanError[],
  rootRel: string,
  allowEscape: string[] | undefined,
): { real: string; stat: fs.Stats } | undefined {
  let lst: fs.Stats;
  try {
    lst = fs.lstatSync(abs);
  } catch {
    return undefined; // 不存在 = 此处无文件，不报错
  }

  if (!lst.isSymbolicLink()) {
    return { real: abs, stat: lst };
  }

  const real = tryRealpath(abs);
  if (real === undefined) return undefined; // 悬空链接 → 静默跳过（allowlist 命中也不例外）
  if (!isWithinRoot(real, rootReal) && !isEscapeAllowed(rootRel, allowEscape)) {
    errors.push({ path: abs, message: `symlink 逃逸 adapter root: ${real}` });
    return undefined;
  }
  try {
    return { real, stat: fs.statSync(real) };
  } catch {
    return undefined;
  }
}

function scanCategory(
  root: string,
  rootReal: string,
  category: string,
  cfg: CategoryConfig,
  ignore: string[] | undefined,
  allowEscape: string[] | undefined,
  errors: ScanError[],
): CategorySnapshot {
  const files: SnapshotFiles = new Map<string, SnapshotEntry>();

  for (const p of cfg.paths) {
    if (isDirPath(p)) {
      const relDir = p.replace(/\/+$/, '');
      if (relDir === '') continue;
      if (isIgnored(relDir, ignore) || isExcluded(relDir, cfg.exclude)) continue;
      const absDir = path.join(root, relDir);
      const resolved = resolveConfiguredPath(absDir, rootReal, errors, relDir, allowEscape);
      // 目录不存在 / 不是目录 / symlink 逃逸 → 该分类此处无文件
      if (resolved === undefined || !resolved.stat.isDirectory()) continue;

      // 合法（root 内）的 symlink 目录 → 用真实路径作扫描起点；walk 内部的 visited 集合
      // 仍会截断回环，readEntry 也用 realpath 读（scanBase 已 realpath）。
      const scanBase = resolved.real;
      const collected: string[] = [];
      walk(scanBase, '', 0, {
        ignore,
        exclude: cfg.exclude,
        allowEscape,
        errors,
        out: collected,
        rootReal,
        // allowlist 匹配基准用**声明的 path**（而非 realpath），与用户写的 glob 对齐
        rootRelBase: relDir,
        visited: new Set<string>(),
      });
      for (const rel of collected) {
        const entry = readEntry(path.join(scanBase, rel), cfg.mode, errors);
        if (entry) files.set(normalizeRel(rel), entry);
      }
    } else {
      if (isIgnored(p, ignore) || isExcluded(p, cfg.exclude)) continue;
      const abs = path.join(root, p);
      const resolved = resolveConfiguredPath(abs, rootReal, errors, p, allowEscape);
      if (resolved === undefined || !resolved.stat.isFile()) continue;
      const entry = readEntry(resolved.real, cfg.mode, errors);
      // §1.6：单文件 → relPath = 文件名本身
      if (entry) files.set(basenameOf(normalizeRel(p)), entry);
    }
  }

  // 稳定顺序：按 relPath 排序，方便测试与 diff（walk 内的中间排序已删除，这里是唯一来源）
  const sorted: SnapshotFiles = new Map<string, SnapshotEntry>();
  for (const key of [...files.keys()].sort()) {
    const value = files.get(key);
    if (value) sorted.set(key, value);
  }

  return { adapterId: '', category, mode: cfg.mode, files: sorted };
}

/**
 * 只读扫描一个 adapter root，产出 AdapterSnapshot。
 * root 不存在 → 空 categories + errors，不 throw。
 */
export function scanAdapter(adapterId: string, config: AdapterConfig): ScanOutcome {
  const errors: ScanError[] = [];
  const root = expandHome(config.root);

  const snapshot: AdapterSnapshot = { adapterId, categories: [] };

  if (config.enabled === false) {
    return { snapshot, errors };
  }

  let rootStat: fs.Stats;
  try {
    rootStat = fs.statSync(root);
  } catch (err) {
    errors.push({ path: root, message: (err as Error).message });
    return { snapshot, errors };
  }
  if (!rootStat.isDirectory()) {
    errors.push({ path: root, message: 'not a directory' });
    return { snapshot, errors };
  }

  const rootReal = tryRealpath(root) ?? root;

  for (const [category, cfg] of Object.entries(config.categories)) {
    if (cfg.enabled === false) continue;
    const cat = scanCategory(root, rootReal, category, cfg, config.ignore, config.allowEscape, errors);
    cat.adapterId = adapterId;
    snapshot.categories.push(cat);
  }

  return { snapshot, errors };
}

File: internal/adapter/scan.go
package adapter

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/zzjcool/homer-cli/internal/core"
)

// ScanError is a non-fatal diagnostic emitted while collecting a snapshot.
// A missing configured file is not an error; errors are reserved for an
// unreadable root/path, a failed directory read, or a denied symlink escape.
type ScanError struct {
	Path    string
	Message string
}

// ScanOutcome contains the snapshot and diagnostics from one adapter scan.
type ScanOutcome struct {
	Snapshot core.AdapterSnapshot
	Errors   []ScanError
}

const maxWalkDepth = 32

// scanState is per category.  A category has its own visited set so that a
// directory declared by two categories is independently represented in each
// category, while a symlink loop inside one category is cut off.
type scanState struct {
	Ignore      []string
	Exclude     []string
	AllowEscape []string
	Errors      *[]ScanError
	Files       *[]string
	RootReal    string
	RootRelBase string
	Visited     map[string]struct{}
}

func expandHome(input string) string {
	if input != "~" && !strings.HasPrefix(input, "~/") && !strings.HasPrefix(input, `~\`) {
		return input
	}

	home := os.Getenv("HOME")
	if home == "" {
		if resolved, err := os.UserHomeDir(); err == nil {
			home = resolved
		}
	}
	if home == "" {
		return input
	}
	if input == "~" {
		return home
	}
	return filepath.Join(home, filepath.FromSlash(input[2:]))
}

func isDirPath(path string) bool { return strings.HasSuffix(path, "/") }

func normalizeRel(path string) string {
	return filepath.ToSlash(path)
}

func basenameOf(rel string) string {
	rel = normalizeRel(rel)
	if idx := strings.LastIndexByte(rel, '/'); idx >= 0 {
		return rel[idx+1:]
	}
	return rel
}

func joinRel(rootRelBase, rel string) string {
	rootRelBase = strings.Trim(rootRelBase, "/")
	rel = strings.Trim(rel, "/")
	switch {
	case rootRelBase == "":
		return rel
	case rel == "":
		return rootRelBase
	default:
		return rootRelBase + "/" + rel
	}
}

func isIgnored(rel string, ignore []string) bool {
	return MatchesIgnore(normalizeRel(rel), ignore)
}

// isExcluded preserves the TypeScript category-exclude rule: a pattern is
// tried against the whole category-relative path, its basename, and every
// path segment.  This makes *cache* prune both cache-like files and entire
// cache-like directories while adapter-level ignore remains path-strict.
func isExcluded(rel string, exclude []string) bool {
	if len(exclude) == 0 {
		return false
	}
	rel = normalizeRel(rel)
	if MatchesIgnore(rel, exclude) || MatchesIgnore(basenameOf(rel), exclude) {
		return true
	}
	for _, segment := range strings.Split(rel, "/") {
		if segment != "" && MatchesIgnore(segment, exclude) {
			return true
		}
	}
	return false
}

func isEscapeAllowed(rootRel string, allowEscape []string) bool {
	for _, pattern := range allowEscape {
		// Config validation owns the user-facing error, but the scanner also
		// keeps an unsafe value fail-closed when called directly (for example
		// by an embedding caller that constructed AdapterConfig by hand).
		if IsBareGlob(pattern) {
			continue
		}
		if GlobMatch(rootRel, pattern) {
			return true
		}
	}
	return false
}

func tryRealpath(path string) (string, bool) {
	real, err := filepath.EvalSymlinks(path)
	if err != nil {
		return "", false
	}
	return real, true
}

func isWithinRoot(child, root string) bool {
	child = filepath.Clean(child)
	root = filepath.Clean(root)
	if child == root {
		return true
	}
	// Keep an existing volume/root separator (for example "/" or "C:\\")
	// instead of appending a second one.
	prefix := root
	if !strings.HasSuffix(prefix, string(os.PathSeparator)) {
		prefix += string(os.PathSeparator)
	}
	return strings.HasPrefix(child, prefix)
}

func addScanError(errors *[]ScanError, path, message string) {
	*errors = append(*errors, ScanError{Path: path, Message: message})
}

// visitSymlink applies the containment policy before following a symlink.
// Dangling links are intentionally silent.  A denied escape is diagnosable,
// but never contributes target contents to the snapshot.
func visitSymlink(abs, rel string, depth int, state *scanState) {
	real, ok := tryRealpath(abs)
	if !ok {
		return
	}

	rootRel := joinRel(state.RootRelBase, normalizeRel(rel))
	if !isWithinRoot(real, state.RootReal) && !isEscapeAllowed(rootRel, state.AllowEscape) {
		addScanError(state.Errors, abs, fmt.Sprintf("symlink 逃逸 adapter root: %s", real))
		return
	}

	info, err := os.Stat(real)
	if err != nil {
		// A target can disappear between realpath and stat.  Treat it as a
		// dangling path, matching the scan's best-effort semantics.
		return
	}
	if info.IsDir() {
		walk(abs, rel, depth+1, state)
	} else if info.Mode().IsRegular() {
		*state.Files = append(*state.Files, normalizeRel(rel))
	}
}

func walk(absBase, relBase string, depth int, state *scanState) {
	if depth > maxWalkDepth {
		return
	}

	realBase, ok := tryRealpath(absBase)
	if !ok {
		return
	}
	if _, seen := state.Visited[realBase]; seen {
		return
	}
	state.Visited[realBase] = struct{}{}

	entries, err := os.ReadDir(absBase)
	if err != nil {
		addScanError(state.Errors, absBase, err.Error())
		return
	}

	for _, entry := range entries {
		name := entry.Name()
		rel := name
		if relBase != "" {
			rel = relBase + "/" + name
		}
		rel = normalizeRel(rel)
		if isIgnored(rel, state.Ignore) || isExcluded(rel, state.Exclude) {
			continue
		}

		abs := filepath.Join(absBase, filepath.FromSlash(name))
		if entry.Type()&os.ModeSymlink != 0 {
			visitSymlink(abs, rel, depth, state)
			continue
		}
		if entry.IsDir() {
			walk(abs, rel, depth+1, state)
			continue
		}
		if entry.Type().IsRegular() {
			*state.Files = append(*state.Files, rel)
		}
	}
}

func entryKind(mode core.SyncMode, content string) string {
	return core.EntryKindFor(mode, content)
}

func readEntry(abs string, mode core.SyncMode, errors *[]ScanError) (core.SnapshotEntry, bool) {
	content, err := os.ReadFile(abs)
	if err != nil {
		addScanError(errors, abs, err.Error())
		return core.SnapshotEntry{}, false
	}
	return core.SnapshotEntry{Kind: entryKind(mode, string(content)), Content: string(content)}, true
}

type configuredPath struct {
	Real string
	Info os.FileInfo
}

// resolveConfiguredPath handles a symlink when the declared category path
// itself is a link.  A missing path and a dangling link both mean "no file";
// an escape is the one case that is reported.
func resolveConfiguredPath(abs, rootReal, rootRel string, allowEscape []string, errors *[]ScanError) (configuredPath, bool) {
	// lstat first distinguishes a genuinely absent path from a path that
	// races with deletion.  EvalSymlinks then resolves the entire path, not
	// just its final component, so a configured "dir/file" cannot smuggle an
	// outside file through an escaping symlink in dir.
	if _, err := os.Lstat(abs); err != nil {
		return configuredPath{}, false
	}
	real, ok := tryRealpath(abs)
	if !ok {
		return configuredPath{}, false
	}
	if !isWithinRoot(real, rootReal) && !isEscapeAllowed(rootRel, allowEscape) {
		addScanError(errors, abs, fmt.Sprintf("symlink 逃逸 adapter root: %s", real))
		return configuredPath{}, false
	}
	info, err := os.Stat(real)
	if err != nil {
		return configuredPath{}, false
	}
	return configuredPath{Real: real, Info: info}, true
}

func scanCategory(root, rootReal, category string, cfg core.CategoryConfig, ignore, allowEscape []string, errors *[]ScanError) core.CategorySnapshot {
	files := make(core.SnapshotFiles)

	for _, declared := range cfg.Paths {
		declared = normalizeRel(declared)
		if isDirPath(declared) {
			relDir := strings.TrimRight(declared, "/")
			if relDir == "" || isIgnored(relDir, ignore) || isExcluded(relDir, cfg.Exclude) {
				continue
			}

			absDir := filepath.Join(root, filepath.FromSlash(relDir))
			resolved, ok := resolveConfiguredPath(absDir, rootReal, relDir, allowEscape, errors)
			if !ok || !resolved.Info.IsDir() {
				continue
			}

			collected := make([]string, 0)
			state := scanState{
				Ignore:      ignore,
				Exclude:     cfg.Exclude,
				AllowEscape: allowEscape,
				Errors:      errors,
				Files:       &collected,
				RootReal:    rootReal,
				RootRelBase: relDir,
				Visited:     make(map[string]struct{}),
			}
			walk(resolved.Real, "", 0, &state)
			for _, rel := range collected {
				entry, ok := readEntry(filepath.Join(resolved.Real, filepath.FromSlash(rel)), cfg.Mode, errors)
				if ok {
					files[normalizeRel(rel)] = entry
				}
			}
			continue
		}

		if isIgnored(declared, ignore) || isExcluded(declared, cfg.Exclude) {
			continue
		}
		abs := filepath.Join(root, filepath.FromSlash(declared))
		resolved, ok := resolveConfiguredPath(abs, rootReal, declared, allowEscape, errors)
		if !ok || !resolved.Info.Mode().IsRegular() {
			continue
		}
		entry, ok := readEntry(resolved.Real, cfg.Mode, errors)
		if ok {
			// A single-file category is keyed by basename, not by the
			// declaration's directory prefix.
			files[basenameOf(declared)] = entry
		}
	}

	return core.CategorySnapshot{
		AdapterID: "",
		Category:  category,
		Mode:      cfg.Mode,
		Files:     files,
	}
}

// categoryOrder makes snapshots deterministic despite CategoryConfig being a
// map in the frozen Go interface.  For the three built-in adapters it follows
// the insertion order of the frozen TypeScript defaults; custom categories are
// appended in lexical order.
func categoryOrder(adapterID string, categories map[string]core.CategoryConfig) []string {
	preferred := map[string][]string{
		"pi":       {"settings", "skills", "extensions", "agents", "models", "prompts", "themes"},
		"herdr":    {"config"},
		"opencode": {"config", "plugins", "locks"},
	}[adapterID]

	ordered := make([]string, 0, len(categories))
	seen := make(map[string]struct{}, len(categories))
	for _, name := range preferred {
		if _, ok := categories[name]; ok {
			ordered = append(ordered, name)
			seen[name] = struct{}{}
		}
	}
	rest := make([]string, 0, len(categories)-len(ordered))
	for name := range categories {
		if _, ok := seen[name]; !ok {
			rest = append(rest, name)
		}
	}
	sort.Strings(rest)
	return append(ordered, rest...)
}

// ScanAdapter reads one adapter root without writing to the tool directory.
// Missing roots produce an empty snapshot and one diagnostic; missing
// configured files/directories produce empty category entries without errors.
func ScanAdapter(adapterID string, config core.AdapterConfig) ScanOutcome {
	outcome := ScanOutcome{
		Snapshot: core.AdapterSnapshot{AdapterID: adapterID, Categories: []core.CategorySnapshot{}},
		Errors:   make([]ScanError, 0),
	}
	if config.Enabled != nil && !*config.Enabled {
		return outcome
	}

	root := expandHome(config.Root)
	rootInfo, err := os.Stat(root)
	if err != nil {
		addScanError(&outcome.Errors, root, err.Error())
		outcome.Snapshot.Categories = []core.CategorySnapshot{}
		return outcome
	}
	if !rootInfo.IsDir() {
		addScanError(&outcome.Errors, root, "not a directory")
		outcome.Snapshot.Categories = []core.CategorySnapshot{}
		return outcome
	}

	rootReal, ok := tryRealpath(root)
	if !ok {
		rootReal = root
	}

	for _, category := range categoryOrder(adapterID, config.Categories) {
		cfg := config.Categories[category]
		if cfg.Enabled != nil && !*cfg.Enabled {
			continue
		}
		cat := scanCategory(root, rootReal, category, cfg, config.Ignore, config.AllowEscape, &outcome.Errors)
		cat.AdapterID = adapterID
		outcome.Snapshot.Categories = append(outcome.Snapshot.Categories, cat)
	}
	return outcome
}

// Lower-case aliases are useful to tests kept in package adapter and mirror
// the original TypeScript naming without duplicating implementation.
func scanAdapter(adapterID string, config core.AdapterConfig) ScanOutcome {
	return ScanAdapter(adapterID, config)
}
